GitHub checks functionality

jamestoyer opened the pull request:
This integrates the ability to use the GitHub Checks API to post the results of the policies. This means the results will be included in the GH UI instead of jumping to a different UI, although you can still jump to the old view at the bottom of the checks page.

Here is an example of what this would look like on the PR:

![screen shot 2018-12-21 at 16 16 43](https://user-images.githubusercontent.com/380725/50766825-99f43180-1272-11e9-886a-9a9f315bd100.png)

And on the checks page:
 
![screen shot 2018-12-21 at 14 45 39](https://user-images.githubusercontent.com/380725/50766833-a2e50300-1272-11e9-894a-46e02c1f5bac.png)

Replaces #45

jamestoyer committed:
GitHub checks created on first PR request

This only creates a Check Run when there are no other Check Runs. It
also currently combines all the results under one check and doesn't
split them out.

jamestoyer committed:
Existing check run checks removed

The check to see if there are existing check runs has been removed. When
you have a check that has completed you don't need to updated, just
creating a new one will update the status.

jamestoyer committed:
Results of evaluation processed to generate checks

Instead of having one high level check run there is a check run for
approvals and disapprovals

jamestoyer committed:
Children of policy results rendered

Each result can have N number of child results. This commit will render
them out recursively

jamestoyer committed:
Action type changed for pending

When "Action Required" was set for a check it would link off to the
external Policy Bot site. Changing it to fail should stop this

jamestoyer committed:
Configuration flags for check runs added

GitHub statuses and GitHub check runs can now be enabled/disabled using
configuration flags allowing the toggling of those features. The example
config file has also been updated to reflect this

jamestoyer committed:
Handler created for the Check Run event type

This handler allows for a manual refresh of the checks in the case they
might have failed or need a refresh

jamestoyer committed:
Check run info included in README

README has been updated to include information about the settings
required for Checks

jamestoyer committed:
Go fmt fixes

jamestoyer committed:
Installation ID name in check_run to pass tests

Installation ID was incorrectly named so tests were not passing. This
has been rectified so `d` is now `D`

jamestoyer committed:
String builder outputs explicitly defined

By default the builder `WriteString` method returns an error. This
requires the error being sent to a variable to ensure it passes checks,
regardless of the fact for a string builder `nil` is returned

jamestoyer committed:
Check suite event handler created

The payload for the check run and check suite events are different. This
adds handling for the re-request of the running of an entire check suite

jamestoyer committed:
Context added to check names

The name of the GitHub app and the branch have been added to the name of
the run. This helps give extra context to the checks that have been run.

jamestoyer commented:
I've added contextual information to the name of the check run, including the name of the Policy Bot and then name of branch

bluekeyes commented on server/handler/base.go:
```diff
@@ -58,6 +59,16 @@ type PullEvaluationOptions struct {
+	// EnableGitHubCheckRuns enables posting of the results of the policy evaluations as GitHub checks. This means that
+	// that full results are visible on the checks tab of a GitHub pull request. By default this is turned off as the
+	// `checks:write` permission is required for the GitHub application
+	EnableGitHubCheckRuns bool `yaml:"enable_github_check_runs"`
```
I like that this is configurable (and that you can disable the "old" method if necessary), but it's unfortunate you can configure the server to not post any results. Is there an alternate configuration that could avoid this, like a single option that takes 3 values? Or validation on startup?

bluekeyes commented on server/handler/base.go:
```diff
@@ -111,6 +122,87 @@ func (b *Base) postGitHubRepoStatus(ctx context.Context, client *github.Client,
+	case "unknown":
+		conclusion = "failure"
+	case "pending":
+		conclusion = "failure"
```
Rather than `pending` resulting in `completed` with a `failure` conclusion, does it make more sense to have the run remain in the `queued` or `in_progress` state in this case? That better matches the old commit statuses and suggests that the app is still waiting for something to happen.

jamestoyer replied:
I've changed `pending` to have a `action_required` conclusion. This better follows the checks workflow whereby a check that has been run must have a completed status when it is not in progress. It also makes sure that an "update" isn't needed on a check, thus ensuring we don't (currently) need to store state in the bot.

Below is the workflow suggested by [GitHub](https://developer.github.com/v3/checks/runs/)
![](https://developer.github.com/assets/images/check_runs.png)

jamestoyer replied:
Just to add, this does now mean there is no manual way to force the checks to re-run, it will always require an update to the rest of the PR. 

In the example below, the first check run has a status of `action_required` and the second has one of `failure`

![image](https://user-images.githubusercontent.com/380725/50839865-2aa53d00-1359-11e9-99ee-c40adfcbca75.png)

bluekeyes commented on server/handler/base.go:
```diff
@@ -167,6 +259,15 @@ func (b *Base) EvaluateFetchedConfig(ctx context.Context, mbrCtx pull.Membership
 		return errors.Errorf("evaluation resulted in unexpected state: %s", result.Status)
 	}
 
-	err = b.PostStatus(ctx, client, pr, statusState, statusDescription)
```
Note that there are other locations in the code that call `PostStatus`; those probably need to be updated to handle check runs as well.

bluekeyes commented on server/handler/check_run.go:
```diff
@@ -0,0 +1,66 @@
+
+			// HACK: This gets around a lack of context from the PR associated with a check run. As the API might
+			// change later, this should be re-evaluated at a later date
+			pullRequest.Base.Repo = event.GetRepo()
```
It might be safer to make an additional API call here to load the full PR object. As is, I could see us adding other code that uses missing parts of this object and miss it during testing, particularly if check runs are disabled.

bluekeyes commented on server/handler/check_suite.go:
```diff
@@ -0,0 +1,66 @@
+	}
+
+	switch event.GetAction() {
+	case "rerequested":
```
Does this also need to handle the `requested` event?

jamestoyer replied:
The `check_run` and `check_suite` events exist to capture when a user manually re-requests a suite/run to re-run. The other statuses would actually be triggered by the initial creation of a run/suite which are done automatically when posting a new run is created or a new suite is created. So in this case `rerequested` is the only event we care about

See:
- https://developer.github.com/v3/activity/events/types/#checksuiteevent
- https://developer.github.com/v3/activity/events/types/#checkrunevent

bluekeyes commented on server/handler/base.go:
```diff
@@ -111,6 +122,87 @@ func (b *Base) postGitHubRepoStatus(ctx context.Context, client *github.Client,
+	indentValue := strings.Repeat(" ", 2*level)
+
+	for _, childResult := range results {
+		_, _ = summaryBuilder.WriteString("\n")
```
style: you shouldn't have to ignore these return values... the latest version of `errcheck` includes exceptions for `strings.Builder` methods, but unfortunately the Godel asset containing that hasn't been released yet.

For now, you can add excludes to the `godel/config/checks.yml` file:

```yaml
checks:
  errcheck:
    config:
      exclude:
        - (*strings.Builder).Write
        - (*strings.Builder).WriteByte
        - (*strings.Builder).WriteRune
        - (*strings.Builder).WriteString
```

jamestoyer replied:
Updated the Godel config to ignore `strings.Builder` and revoked the previous changes to make this look a little less cluttered

bluekeyes commented on server/handler/base.go:
```diff
@@ -111,6 +122,87 @@ func (b *Base) postGitHubRepoStatus(ctx context.Context, client *github.Client,
+		_, _ = summaryBuilder.WriteString(childSummaries)
+	}
+
+	return summaryBuilder.String()
```
Did you look at using `text/template` for this, similar to how the UI version is rendered? I'm not sure if it would be simpler, but a recursive template might be easier to understand than the string formatting indentation here.

bluekeyes commented on server/handler/base.go:
```diff
@@ -111,6 +122,87 @@ func (b *Base) postGitHubRepoStatus(ctx context.Context, client *github.Client,
+	repo := pr.GetBase().GetRepo().GetName()
+
+	for _, result := range evaluationResult.Children {
+		if err := b.createGitHubRepoCheck(ctx, client, pr, owner, repo, sha, *result); err != nil {
```
This ends up creating both approval and disapproval runs, right? Would it be better to combine them into a single run or does the fact that they are part of the same suite effectively combine them?

jamestoyer replied:
That's correct. In my mind it more clearly splits out what has been the failed checks and the checks that have not. However, as they are both part of the same suite, which is tied to a ref, it shouldn't really matter underneath

bluekeyes reviewed (commented):
Overall, this looks good. I'll have to do some research into how check runs are added to branch protection to make sure that part still works.

jamestoyer committed:
String builder return checks ignored by godel

The godel configuration for the check plugin has been updated to ignore
not assigning the return error to a variable. This is because it always
returns nil.

This revokes the changes made by 820d85535dc001b50c1e597527db73b92187af7d

jamestoyer committed:
Pending statuses updated to return action required

Checks that have a pending status no longer return failure but instead
action_required

jamestoyer commented:
@bluekeyes With branch permissions it works exactly the same as traditional statuses do. You should be able to create protections in the same way you do normal statuses

jamestoyer committed:
Check suite handler updated to load PRs explicitly

Previously there was some hacking and brute forcing to make sure the PR
object on a check suite event matched a full PR. This instead loads the
full PR from the GitHub API to make sure it's correct.

jamestoyer committed:
Re-run checks for suite refactored

This refactors out running the checks for a suite out into the base
handler class as it will be used in multiple handlers

jamestoyer committed:
Check run handler updated to use base processor

The check run handler has been updated to use the refactored processing
method for events that fire a rerequest

NargiT commented:
is this dead ?

wrslatz commented:
Looks like a new issue for this has been opened: https://github.com/palantir/policy-bot/issues/348
